core/fragment.c

```diff
@@ -1139,6 +1139,18 @@ hashtable_fragment_reset(dcontext_t *dcontext, fragment_table_t *table)
         if (!dynamo_exited && !dynamo_resetting)
             ASSERT_TABLE_SYNCHRONIZED(table, WRITE);
     });
+#    if !defined(DEBUG) && defined(CLIENT_INTERFACE)
+    if (!dr_fragment_deleted_hook_exists())
+        return;
+    /* i#4226: Avoid the slow deletion code and just invoke the event. */
+    for (i = 0; i < table->capacity; i++) {
+        f = table->table[i];
+        if (!REAL_FRAGMENT(f))
+            continue;
+        instrument_fragment_deleted(dcontext, f->tag, f->flags);
+    }
+    return;
+#    endif
     /* Go in reverse order (for efficiency) since using
```

Comment:
Most likely yes, but I am wondering whether or not we need to keep the remaining debug code given that it is not reachable anymore on release?

Reply:
I'm not following: the debug code is executed in debug build and helps with finding memory leaks, so I'm not seeing why it would be removed?

Reply:
Thanks for clearing that up. Just wanted to make sure that the debug code is *still* relevant given the quick return now present on release builds.

Reply:
Note that in release builds without any fragment deletion event, this function is not even called at all.  Most of the individual cleanup is skipped and we just munmap from a higher level (like an arena dealloc).